model/delivery/AutoStartProctorService.php

```diff
@@ -0,0 +1,107 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2022 (original work) Open Assessment Technologies SA;
+ *
+ */
+
+declare(strict_types=1);
+
+namespace oat\ltiProctoring\model\delivery;
+
+use common_session_Session as Session;
+use oat\oatbox\log\LoggerService;
+use oat\oatbox\user\User;
+use oat\taoDelivery\model\execution\DeliveryExecution;
+use oat\taoDelivery\model\execution\DeliveryExecutionInterface;
+use oat\taoLti\models\classes\LtiException;
+use oat\taoLti\models\classes\TaoLtiSession;
+use oat\taoLti\models\classes\user\LtiUser;
+use oat\taoProctoring\model\execution\DeliveryExecution as ProctoredDeliveryExecution;
+use oat\taoQtiTest\models\TestSessionService;
+
+class AutoStartProctorService
+{
+    private const CUSTOM_LTI_AUTOSTART = 'custom_autostart';
+
+    /**
+     * @var TestSessionService
+     */
+    private $testSessionService;
+
+    /**
+     * @var LoggerService
+     */
+    private $logger;
+
+    public function __construct(TestSessionService $testSessionService, LoggerService $logger)
+    {
+        $this->testSessionService = $testSessionService;
+        $this->logger = $logger;
+    }
+
+    public function execute(DeliveryExecution $deliveryExecution, User $user, Session $session): ?string
+    {
+        if (false === $this->validateAutoStart($deliveryExecution, $user)) {
+            return null;
+        }
+
+        $deliveryExecution->getImplementation()->setState(ProctoredDeliveryExecution::STATE_AUTHORIZED);
+
+        return $this->getUrlRunDeliveryExecution($deliveryExecution, $session);
+    }
+
+    private function validateAutoStart(DeliveryExecution $deliveryExecution, User $user): bool
+    {
+        if (null === $this->testSessionService->getTestSession($deliveryExecution)) {
+            return $this->isAutostartEnabled($user);
+        }
+
+        return false;
+    }
+
+    private function isAutostartEnabled(User $user): bool
+    {
+        if (false === $user instanceof LtiUser) {
+            return false;
+        }
+
+        try {
+            $ltiLaunchData = $user->getLaunchData();
+            if ($ltiLaunchData->hasVariable(self::CUSTOM_LTI_AUTOSTART)) {
+                return $ltiLaunchData->getBooleanVariable(self::CUSTOM_LTI_AUTOSTART);
+            }
+        } catch (LtiException $exception) {
+            $this->logger->warning(
+                "Invalid custom LTI parameter",
+                ['message' => $exception->getMessage(), 'file' => $exception->getFile(), 'line' => $exception->getLine()]
+            );
+        }
+
+        return false;
+    }
+
+    private function getUrlRunDeliveryExecution(DeliveryExecutionInterface $deliveryExecution, Session $session): string
+    {
+        return _url(
+            'runDeliveryExecution',
+            'DeliveryServer',
+            $session instanceof TaoLtiSession ? 'ltiProctoring' : 'taoProctoring',
```

Comment:
Well, this code block can only be reached if the user is `LtiUser` and `LtiLaunchData` has a specific `autostart=true` parameter, so we are kind of sure that if a redirect is necessary then it's to `ltiProctoring`.
And maybe it's better fitting to build this url and redirect still in the controller?